### ISSUE TITLE
better error reporting for cradle failure

### ISSUE THREAD

**Issue by alanz** [haskell/haskell-language-server#1245]
I don't have stack set up on my machine, did a clean checkout of this repo, configured and built with cabal, and when I try to open `../ghcide/test/src/Development/IDE/Test.hs` I get told the language server exited, and the following shows up in the `*lsp-haskell::stderr*` buffer.

```
Current directory: /Users/alanzimm/repos/haskell/haskell-language-server
Operating system: darwin
Arguments: ["--lsp","-d","-l","/var/folders/vy/nf8sc07944324wf17pg48m2r0000gn/T/hls.log"]
Cradle directory: /Users/alanzimm/repos/haskell/haskell-language-server
Cradle type: Stack

Tool versions found on the $PATH
cabal:		3.2.0.0
stack:		2.1.3
ghc:		8.10.3


Consulting the cradle to get project GHC version...
Failed to get project GHC version:CradleError {cradleErrorDependencies = [], cradleErrorExitCode = ExitFailure 1, cradleErrorStderr = ["Error when calling stack exec --silent ghc -- --numeric-version","",""]}

Process lsp-haskell stderr<3> finished
```

### Your environment
Output of `haskell-language-server --probe-tools` or `haskell-language-server-wrapper --probe-tools`:

```sh
haskell-language-server --probe-tools
haskell-language-server version: 0.5.1.0 (GHC: 8.10.2) (PATH: /Users/alanzimm/.cabal/store/ghc-8.10.2/hskll-lngg-srvr-0.5.1.0-9b087dc4/bin/haskell-language-server)
Tool versions found on the $PATH
cabal:		3.2.0.0
stack:		2.1.3
ghc:		8.10.3
```
Which lsp-client do you use: emacs lsp-mode

Describe your project (alternative: link to the project):

Contents of `hie.yaml`: none

### Steps to reproduce
Make a clean checkout of this repo at f03a7fa55d909c864771f0989abace8793b5b15a, build it, and open the designated file

### Expected behaviour
It should start up properly

### Actual behaviour
I get `Server lsp-haskell:793 status:starting exited with status exit. Do you want to restart it? (y or n)`, and the log at the top

**Comment by jneira** [comment]
There is an old issue from ghcide asking for a better error reporting when failing builds through hie-bios: #710 
In that case the server (ghcide) did not crash afaiu.

**Comment by fendor** [comment]
I opened https://github.com/mpickering/hie-bios/pull/254 to tackle this, but we didn't verify yet that it has no unintended side-effects

**Comment by jneira** [comment]
I see lot of reports of this: hls failing to load the project with `stack setup --silent` with no clue in the log about the concrete error.
The rutinary advise is tell users run the command directly to access the log.

**Comment by fendor** [comment]
We should probably capture the output from `stack setup` and write everything into `CradleError`.

**Comment by jneira** [comment]
> 
> 
> We should probably capture the output from `stack setup` and write everything into `CradleError`.

But the ghc setup (downloading and installing) would make it lot lot longer :thinking:

**Comment by fendor** [comment]
True, but otherwise it is rather hard to debug, and we should ideally not require users to look into the logs to understand what went wrong.

**Comment by jneira** [comment]
And what about make two calls one silent and other one with output? or only populate/show the log in `CradleError` if the command exits with error?

**Comment by fendor** [comment]
The latter is what I had in mind, we only want to show the logs if some error is thrown

**Comment by jneira** [comment]
The two calls will remove the ghc installation noise (done in the first one) though

**Comment by fendor** [comment]
Fair enough, but if the error happens during the ghc installation, the work will be done twice.
I think, personally, I'd prefer the longer error message, but also getting feedback (e.g. crashes) sooner. But I totally see why two calls are compelling, too.
